pkg/repository/v1manifest/keys.go

```
@@ -0,0 +1,147 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1manifest
+
+import (
+	"crypto/sha256"
+	"fmt"
+
+	cjson "github.com/gibson042/canonicaljson-go"
+	"github.com/pingcap-incubator/tiup/pkg/repository/crypto"
+	"github.com/pingcap/errors"
+)
+
+// ErrorNotPrivateKey indicate that it need a private key, but the supplied is not.
+var ErrorNotPrivateKey = errors.New("not a private key")
+
+// NewKeyInfo make KeyInfo from private key, public key should be load from json
+func NewKeyInfo(privKey []byte) *KeyInfo {
+	// TODO: support other key type and scheme
+	return &KeyInfo{
+		Algorithms: []string{"sha256", "sha512"},
+		Type:       crypto.KeyTypeRSA,
+		Scheme:     crypto.KeySchemeRSASSAPSSSHA256,
+		Value: map[string]string{
+			"private": string(privKey),
+		},
+	}
+}
+
+// GenKeyInfo generate a new private KeyInfo
+func GenKeyInfo() (*KeyInfo, error) {
+	// TODO: support other key type and scheme
+	_, priv, err := crypto.NewKeyPair(crypto.KeyTypeRSA, crypto.KeySchemeRSASSAPSSSHA256)
+	if err != nil {
+		return nil, err
+	}
+	bytes, err := priv.Serialize()
+	if err != nil {
+		return nil, err
+	}
+	return NewKeyInfo(bytes), nil
+}
+
+// ID returns the hash id of the key
+func (ki *KeyInfo) ID() (string, error) {
+	// TODO: apply keyid_hash_algorithms
+	// Make sure private key and coresponed public key has the same id
+	pk, err := ki.publicKey()
+	if err != nil {
+		return "", err
+	}
+	data, err := pk.Serialize()
+	if err != nil {
+		return "", err
+	}
+	value := map[string]string{
+		"public": string(data),
+	}
+
+	payload, err := cjson.Marshal(KeyInfo{
+		Algorithms: ki.Algorithms,
+		Type:       ki.Type,
+		Scheme:     ki.Scheme,
+		Value:      value,
+	})
+	if err != nil {
+		// XXX: maybe we can assume that the error should always be nil since the KeyInfo struct is valid
```

yes this error can be ignored, it should never be an error